src/Microsoft.AspNet.TestHost/WebSocketClient.cs

```diff
@@ -0,0 +1,121 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.WebSockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNet.WebSockets.Protocol;
+using System.Net.Http;
+using System.Security.Cryptography;
+
+namespace Microsoft.AspNet.TestHost
+{
+    public class WebSocketClient
+    {
+        private TestServer _testServer;
+        internal WebSocketClient(TestServer testServer)
+        {
+            _testServer = testServer;
+            ReceiveBufferSize = 1024 * 16;
+            KeepAliveInterval = TimeSpan.FromMinutes(2);
+            SubProtocols = new List<string>();
+        }
+
+        public IList<string> SubProtocols
+        {
+            get;
+            private set;
+        }
+
+        public TimeSpan KeepAliveInterval
+        {
+            get;
+            set;
+        }
+
+        public int ReceiveBufferSize
+        {
+            get;
+            set;
+        }
+
+        public bool UseZeroMask
+        {
+            get;
+            set;
+        }
+
+        public Action<HttpRequestMessage> ConfigureRequest
+        {
+            get;
+            set;
+        }
+
+        public Action<HttpResponseMessage> InspectResponse
+        {
+            get;
+            set;
+        }
+
+        public async Task<WebSocket> ConnectAsync(Uri uri, CancellationToken cancellationToken)
+        {
+            HttpClient client = _testServer.CreateClient();
```

Comment:
I think the WebSocketClient can bypass the HttpClient entirely. HttpClient doesn't have any concept of WebSockets, so tacking them on is pretty awkward. Instead of creating an HttpRequestMessage here just create an IFeatureCollection & DefaultHttpContext directly (This may be a little easer after https://github.com/aspnet/HttpAbstractions/issues/367).